llama/runner/runner.go

```diff
@@ -293,17 +294,23 @@ func (s *Server) shiftContext(seq *Sequence) {
 }
 
 func flushPending(seq *Sequence) bool {
-	for _, p := range seq.pendingResponses {
-		select {
-		case seq.responses <- p:
-		case <-seq.quit:
-			seq.pendingResponses = []string{}
-			return false
-		}
+	joined := strings.Join(seq.pendingResponses, "")
+	seq.pendingResponses = []string{}
+
+	for !utf8.ValidString(joined) {
+		joined = joined[:len(joined)-1]
```

Comment:
I'm not quite seeing how this doesn't throw away the partial utf character since seq.pendingResponses is reset.

Reply:
The primary place where check and queue partial characters is done is down below at line 513. For most cases, we won't call flushPending until we already have a full unicode character and the check here is a no-op.

There are two cases where we call flushPending with a potentially partial unicode character:
-  We've decided to end the sequence anyways - for example, we hit the character generation limit.
- We have data equal to a max unicode character (4 bytes) and the string is still invalid. This keeps us from queuing things indefinitely if there is a broken character that never gets completed.

In both of these cases, things are either getting sent or discarded so I just start chopping characters off the end until the string is either valid or empty.

The second case (max unicode size) may be a little problematic in rare cases where we have a valid string prefixing an invalid partial piece. As a result, I think I'm going to switch back to the open-coded unicode check that only looks at the end of the string.

I'll also add a comment in flushingPending with some more description.